Add support for immutable variables

JackErb opened the pull request:
Add support for immutable variable using the `let` qualifier in GDScript. This variable must be assigned a value at the declaration site. It can never be modified after the initial assignment.

Note that immutability differs from `const` as it can be applied to any runtime variable, whereas `const` requires that the expression be resolved at compile-time.

Implements and closes https://github.com/godotengine/godot-proposals/issues/820_

This works in the following contexts:
1. local variables.
    * `let value = 10.0`
2. function parameters
    * add optional `let` in function signature to make immutable
    * `func foo(let bar: float)`
3. class level variables
    * works for static variables
    * works with @onready
    * `@onready let parent = $".."`

![image](https://github.com/user-attachments/assets/ec133963-0aed-4db8-b0f5-b4ba9887cd36).

=====

TODO:

- [ ] Reach consensus on syntax.
- [ ] Explore where immutability should be checked at runtime. In the VM, or in `GDScriptInstance::set()`?
- [ ] Condense the tests into shared files.

FUTURE TODO (not in the scope of this change):
- Would be nice to not require an initial assignment at the declaration site, but still guarantee it is set only one time.
     * e.g. could be set individually in each branch of a match statement
     * e.g. could be set once in the constructor or ready()
     * Both these cases would throw an error if we attempt to access the value before it is assigned.

JackErb committed:
Immutable local variable

This immutable local variable is defined with 'let'. It must be assigned a value at its declaration site, and cannot be modified after its initial assignment.

e.g. `let value = 10.0`

tetrapod00 commented:
Is there a proposal for this? I do see https://github.com/godotengine/godot-proposals/issues/820 which mentions `let` but I'm not sure if this exactly implements that proposal.

JackErb commented on modules/gdscript/gdscript_analyzer.cpp:
```diff
@@ -2801,16 +2801,19 @@ void GDScriptAnalyzer::update_dictionary_literal_element_type(GDScriptParser::Di
 void GDScriptAnalyzer::reduce_assignment(GDScriptParser::AssignmentNode *p_assignment) {
 	reduce_expression(p_assignment->assigned_value);
 
-#ifdef DEBUG_ENABLED
```
Remove DEBUG_ENABLED so we track assignments in release.

$REVIEW does this break anything? I should test this works as expected in release builds

JackErb commented:
Addresses this proposal: https://github.com/godotengine/godot-proposals/issues/820

It wasn't clear to me what the consensus was with this proposal, since there was so much back and forth discussion on the thread. I'm new to this process so would appreciate any guidance on how to align this change and the proposal.

ValorZard commented:
Whats the difference between using ``let`` and using ``const``?

hpvb commented:
This does add additional work to all variable assignments, be they immutable or not. This will likely have a negative performance impact on all uses. In most cases it likely will be negligible, but for people doing things like procgen in GDSCript this might be a problem (as they will likely be writing to loads of variables)

Calinou commented:
> This does add additional work to all variable assignments, be they immutable or not. This will likely have a negative performance impact on all uses. In most cases it likely will be negligible, but for people doing things like procgen in GDSCript this might be a problem (as they will likely be writing to loads of variables)

We could skip the checks in release builds to ensure the performance doesn't change in projects exported in release mode.

JackErb committed:
Immutable function parameter

Define an immutable function parameter with an optional `let` in the function signature.

This variable cannot be modified. It allows an optional default value.

e.g.
`func foo(let bar, let alpha = 1.0)`

JackErb committed:
Immutable class variables

You can mark a class level variable as immutable with 'let'. This can also optionally be static. This variable must be assigned an initial value and cannot be modified again. Supports @onready annotation.

Analyzer Check:
* Verifies that we don't attempt to modify an immutable variable within a single script.
Runtime VM Check:
* Verifies at runtime that we don't attempt to modify an immutable variable from an external type.

JackErb commented:
Updated the review. It now supports immutable `locals`, `function parameters`, `class members`, and `static members`. Added unit tests for all error messaging.

geekley commented:
Cool! Does it work with lambdas too? Like adding `let` on lambda parameters and getting error when assigning to a `let` variable of the closure from inside the lambda code?

JackErb commented:
> Cool! Does it work with lambdas too? Like adding `let` on lambda parameters and getting error when assigning to a `let` variable of the closure from inside the lambda code?

Yup, just checked, this works with lambdas :)

dalexeev commented:
There is no consensus on this feature, whether it should be `readonly var`, `let` or some other keyword/annotation. Personally I prefer `let`, but @vnen disagreed as far as I remember. So my suggestions are optional to avoid you doing potentially unnecessary work. I haven't reviewed this in detail, but here are a few things I noticed:

1. The core probably shouldn't be touched at all. Instead of adding `property_is_immutable()` you should check for an immutability flag in `GDScript::_set()` and `GDScriptInstance::set()` (need to add the flag to the `GDScript::MemberInfo` structure instead of `PropertyInfo` usage flag.). `PROPERTY_USAGE_IS_IMMUTABLE` flag might be useful for reflection, but I'd rather not add it, at least for now (since we can have max 64 flags). For documentation purposes we can add `qualifiers` field to `DocData::PropertyDoc` struct (like in `DocData::MethodDoc`). And use it in `editor_help.cpp`, `gdscript_docgen.cpp` etc. (The property setter/getter can't be found via `get_property_list()` either, although they are shown in the documentation.)
2. The compiler should only report an analyzer bug when trying to generate a reassignment of an immutable variable, and the VM probably should not be affected at all (to avoid unnecessary runtime checks). If the analyzer can determine what the assignee is, then it can validate this and either produce an error, or the code is safe and the runtime check is not needed. If the analyzer cannot determine this, then the compiler will generate another opcode (`OPCODE_SET_NAMED`) and the VM will go along a generalized dynamic path, so that the runtime check will be performed in `GDScript`/`GDScriptInstance`.
3. Recently, a small refactoring of the GDScript test system was done, now tracking of multiple analyzer/runtime errors is supported. Please avoid using a large number of files, try to create comprehensive tests.

JackErb commented:
I'll avoid making any major refactors here until we reach consensus on the syntax.

> The core probably shouldn't be touched at all. Instead of adding property_is_immutable() you should check for an immutability flag in GDScript::_set() and GDScriptInstance::set()

This is a good point. I did originally implement this inside of `GDScriptInstance::set()`, but I found that @export variable go through this code path when initializing and hot reloading the game. So when I naively put the immutability check in there, exported properties were no longer set to the editor-defined value. This is why I moved it into the VM layer. It might still make more sense to keep it in `GDScriptInstance::set()`, but we'd need a way for @export variable to be set ignoring the immutability flag.

> Recently, a small refactoring of the GDScript test system was done, now tracking of multiple analyzer/runtime errors is supported. Please avoid using a large number of files, try to create comprehensive tests.

Noted!
